x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plugin/DataNodeRequestSender.java

```diff
@@ -236,11 +244,27 @@ private List<ShardSearchFailure> selectFailures() {
     private void sendOneNodeRequest(TargetShards targetShards, ComputeListener computeListener, NodeRequest request) {
         final ActionListener<List<DriverProfile>> listener = computeListener.acquireCompute();
         sendRequest(request.node, request.shardIds, request.aliasFilters, new NodeListener() {
+
+            private final Set<ShardId> pendingRetries = new HashSet<>();
+
             void onAfter(List<DriverProfile> profiles) {
                 nodePermits.get(request.node).release();
                 if (concurrentRequests != null) {
                     concurrentRequests.release();
                 }
+
+                if (pendingRetries.isEmpty() == false && remainingTargetShardSearchAttempts.getAndDecrement() > 0) {
+                    ongoingTargetShardResolutionAttempts.incrementAndGet();
+                    resolveShards(pendingRetries, computeListener.acquireAvoid().delegateFailure((l, resolutions) -> {
+                        for (var entry : resolutions.entrySet()) {
+                            targetShards.shards.get(entry.getKey()).remainingNodes.addAll(entry.getValue());
+                        }
+                        ongoingTargetShardResolutionAttempts.decrementAndGet();
+                        trySendingRequestsForPendingShards(targetShards, computeListener);
+                        l.onResponse(null);
+                    }));
+                }
```

Comment:
The other branch should be in else?

Reply:
Not necessary. We could make progress without waiting for the moved shard(s) resolution in case there are other shards in queue.